src/spatial/directcrrssa.jl

```diff
@@ -0,0 +1,265 @@
+    num_species = size(starting_state, 1)
+    majumps = ma_jumps
+    if majumps === nothing
+        majumps = MassActionJump(Vector{T}(),
+            Vector{Vector{Pair{Int, Int}}}(),
+            Vector{Vector{Pair{Int, Int}}}())
+    end
+
+    next_jump = SpatialJump{Int}(typemax(Int), typemax(Int), typemax(Int)) #a placeholder
+    next_jump_time = typemax(T)
+    rx_rates = LowHigh(RxRates(num_sites(spatial_system), majumps),
+        RxRates(num_sites(spatial_system), majumps);
+        do_copy = false) # do not copy ma_jumps
+    hop_rates = LowHigh(HopRates(hopping_constants, spatial_system),
+        HopRates(hopping_constants, spatial_system);
+        do_copy = false) # do not copy hopping_constants
+    site_rates_high = zeros(T, num_sites(spatial_system))
+    bd = (bracket_data === nothing) ? BracketData{T, eltype(starting_state)}() :
+         bracket_data
+    u_low_high = LowHigh(starting_state)
+
+    DirectCRRSSAJumpAggregation(next_jump, next_jump_time, end_time, bd, u_low_high,
+        rx_rates, hop_rates,
+        site_rates_high, save_positions, rng, spatial_system;
+        num_specs = num_species, kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::DirectCRRSSAJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    fill_rates_and_get_times!(p, integrator, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+# calculate the next jump / jump time
+function generate_jumps!(p::DirectCRRSSAJumpAggregation, integrator, u, params, t)
+    @unpack rng, rt, site_rates_high, rx_rates, hop_rates, spatial_system = p
+    time_delta = zero(t)
+    while true
+        site = sample(rt, site_rates_high, rng)
+        jump = sample_jump_direct(rx_rates.high, hop_rates.high, site, spatial_system, rng)
+        time_delta += randexp(rng)
+        if accept_jump(p, u, jump)
+            p.next_jump_time = t + time_delta / groupsum(rt)
+            p.next_jump = jump
+            break
+        end
+    end
+    nothing
+end
+
+# execute one jump, changing the system state
+function execute_jumps!(p::DirectCRRSSAJumpAggregation, integrator, u, params, t,
+    affects!)
+    update_state!(p, integrator)
+    update_dependent_rates!(p, integrator, t)
+    nothing
+end
+
+######################## SSA specific helper routines ########################
+# Return true if site is accepted.
+function accept_jump(p, u, jump)
+    if is_hop(p, jump)
+        return accept_hop(p, u, jump)
+    else
+        return accept_rx(p, u, jump)
+    end
+end
+
+function accept_hop(p, u, jump)
+    @unpack hop_rates, spatial_system, rng = p
+    species, site = jump.jidx, jump.src
+    acceptance_threshold = rand(rng) * hop_rate(hop_rates.high, species, site)
+    if hop_rate(hop_rates.low, species, site) > acceptance_threshold
+        return true
+    else
+        # compute the real rate. Could have used hop_rates.high as well.
```

Comment:
Does this mean we are storing two copies of the hopping constants for the low vs high (or is it aliased)?

Reply:
Aliased because `do_copy = false` when initializing `LowHigh(hopping_rates)`. Same with `rx_rates`.